include/solvers/time_solver.h

```diff
@@ -261,12 +280,33 @@ class TimeSolver : public ReferenceCountedObject<TimeSolver>
   { _is_adjoint = _is_adjoint_value; }
 
   /**
-   * Function to return 'last_deltat()', returns system.deltat if
-   * fixed timestep solver is used, completed_deltat if the adaptive
+   * Returns system.deltat if fixed timestep solver is used,
+   * the complete timestep size (sum of all substeps) if the adaptive
    * time solver is used.
    */
-  virtual Real last_complete_deltat();
+  virtual Real last_completed_timestep_size();
+
+  /**
+   * Returns the system.deltat timestep used to get to the current time.
+   * For a uniform timestepper, it will simply return the last used system.deltat, for an
+   * adaptive time solver, it will return the last used substep size.
+   */
+  virtual Real last_completed_deltat();
 
+  /**
+   * The final_time variable can be set by the user or in another library
```

Comment:
Indeed, that is what "public:" means.

But what's important here isn't whether I *can* set this, it's whether I *have to* set this.  As written, this looks like a confusing booby trap.  I can't do adjoint time stepping without it?  Then how did ex5?  I can't do QoI postprocessing without it?  What are all those other examples doing?

I get that you're using this because we're too lazy to come up with a better API yet for QoI functionals that have delta functions in t, but that needs to be explained in the documentation, not just in my chat logs.  And if there are any *other* uses then add them too.  A new user should be able to read this comment and answer every question I asked before, and know exactly when set_final_time() needs to be vs doesn't need to be called.